File: src/backend/trip-service/app/api/v1/drivers.py
from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel

from app.services import driver_service, trip_service

router = APIRouter(prefix="/drivers", tags=["drivers"])

_UPCOMING_STATUSES = {"pending", "notified", "calling", "confirmed"}
_ACTIVE_STATUSES = {"confirmed", "notified", "calling"}


def _phone_variants(phone: str) -> list[str]:
    """Room tokens strip '+'; DynamoDB may store either format."""
    p = phone.strip()
    bare = p.lstrip("+")
    variants = [p]
    if p != bare:
        variants.append(bare)
    if not p.startswith("+"):
        variants.append(f"+{bare}")
    return list(dict.fromkeys(variants))


def _trips_for_driver(phone: str) -> list[dict]:
    seen: set[str] = set()
    trips: list[dict] = []
    for variant in _phone_variants(phone):
        for trip in trip_service.list_by_driver(variant, limit=50):
            tid = trip.get("trip_id")
            if tid and tid not in seen:
                seen.add(tid)
                trips.append(trip)
    return sorted(trips, key=lambda t: t.get("created_at", ""), reverse=True)


class RegisterRequest(BaseModel):
    phone:     str
    name:      str
    fcm_token: str = ""


class UpdateDriverRequest(BaseModel):
    name:      str | None = None
    fcm_token: str | None = None
    status:    str | None = None   # online | offline


@router.post("", status_code=201)
def register(body: RegisterRequest):
    existing = driver_service.get(body.phone)
    if existing:
        # re-registration — update token and name
        return driver_service.update(body.phone, {
            "name": body.name,
            "fcm_token": body.fcm_token,
        })
    return driver_service.register(body.phone, body.name, body.fcm_token)


@router.get("")
def list_drivers():
    return driver_service.list_all()


@router.get("/online")
def list_online_drivers():
    return driver_service.list_online()


@router.get("/{phone}")
def get_driver(phone: str):
    driver = driver_service.get(phone)
    if not driver:
        raise HTTPException(status_code=404, detail="Driver not found")
    return driver


@router.get("/me/trips/upcoming")
def upcoming_trips(x_driver_phone: str = Header(..., alias="X-Driver-Phone")):
    """Trips assigned to the driver that are not finished yet (voice-agent tool)."""
    trips = [
        t for t in _trips_for_driver(x_driver_phone)
        if t.get("status") in _UPCOMING_STATUSES
    ]
    return trips


@router.get("/me/trips/current")
def current_trip(x_driver_phone: str = Header(..., alias="X-Driver-Phone")):
    """Active trip for the driver, if any (voice-agent tool)."""
    for trip in _trips_for_driver(x_driver_phone):
        if trip.get("status") in _ACTIVE_STATUSES:
            return trip
    raise HTTPException(status_code=404, detail="No active trip")


@router.patch("/{phone}")
def update_driver(phone: str, body: UpdateDriverRequest):
    driver = driver_service.get(phone)
    if not driver:
        raise HTTPException(status_code=404, detail="Driver not found")

    fields = {k: v for k, v in body.model_dump().items() if v is not None}
    if not fields:
        return driver

    if "status" in fields and fields["status"] not in ("online", "offline"):
        raise HTTPException(status_code=400, detail="status must be online or offline")

    return driver_service.update(phone, fields)

File: src/backend/trip-service/src/routes/drivers.ts
import { Router, Request, Response } from 'express';
import * as driverService from '../services/driverService';
import * as tripService from '../services/tripService';

const router = Router();

const UPCOMING_STATUSES = new Set(['pending', 'notified', 'calling', 'confirmed']);
const ACTIVE_STATUSES = new Set(['confirmed', 'notified', 'calling']);

function phoneVariants(phone: string): string[] {
  const p = phone.trim();
  const bare = p.replace(/^\+/, '');
  const variants = [p];
  if (p !== bare) variants.push(bare);
  if (!p.startsWith('+')) variants.push(`+${bare}`);
  return [...new Set(variants)];
}

async function tripsForDriver(phone: string) {
  const seen = new Set<string>();
  const trips: Awaited<ReturnType<typeof tripService.listByDriver>> = [];
  for (const variant of phoneVariants(phone)) {
    for (const trip of await tripService.listByDriver(variant, 50)) {
      if (trip.trip_id && !seen.has(trip.trip_id)) {
        seen.add(trip.trip_id);
        trips.push(trip);
      }
    }
  }
  return trips.sort((a, b) => b.created_at.localeCompare(a.created_at));
}

router.post('/', async (req: Request, res: Response) => {
  try {
    const { phone, name, fcm_token = '' } = req.body;
    const existing = await driverService.get(phone);
    if (existing) {
      return res.status(201).json(await driverService.update(phone, { name, fcm_token }));
    }
    return res.status(201).json(await driverService.register(phone, name, fcm_token));
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/', async (_req: Request, res: Response) => {
  try {
    return res.json(await driverService.listAll());
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/online', async (_req: Request, res: Response) => {
  try {
    const all = await driverService.listAll();
    return res.json(driverService.listOnline(all));
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/me/trips/upcoming', async (req: Request, res: Response) => {
  try {
    const driverPhone = req.headers['x-driver-phone'] as string;
    if (!driverPhone) return res.status(400).json({ detail: 'X-Driver-Phone header required' });
    const trips = (await tripsForDriver(driverPhone)).filter(t => UPCOMING_STATUSES.has(t.status));
    return res.json(trips);
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/me/trips/current', async (req: Request, res: Response) => {
  try {
    const driverPhone = req.headers['x-driver-phone'] as string;
    if (!driverPhone) return res.status(400).json({ detail: 'X-Driver-Phone header required' });
    const trips = await tripsForDriver(driverPhone);
    const active = trips.find(t => ACTIVE_STATUSES.has(t.status));
    if (!active) return res.status(404).json({ detail: 'No active trip' });
    return res.json(active);
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/:phone', async (req: Request, res: Response) => {
  try {
    const driver = await driverService.get(req.params.phone);
    if (!driver) return res.status(404).json({ detail: 'Driver not found' });
    return res.json(driver);
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.patch('/:phone', async (req: Request, res: Response) => {
  try {
    const driver = await driverService.get(req.params.phone);
    if (!driver) return res.status(404).json({ detail: 'Driver not found' });

    const fields: Record<string, unknown> = {};
    for (const k of ['name', 'fcm_token', 'status']) {
      if (req.body[k] != null) fields[k] = req.body[k];
    }
    if (!Object.keys(fields).length) return res.json(driver);
    if (fields.status && !['online', 'offline'].includes(fields.status as string)) {
      return res.status(400).json({ detail: 'status must be online or offline' });
    }
    return res.json(await driverService.update(req.params.phone, fields));
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

export default router;
